rosbridge_server/src/rosbridge_server/websocket_handler.py

```diff
@@ -65,6 +68,51 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+class IncomingQueue(threading.Thread):
+    """Decouples incoming messages from the Tornado thread.
+
+    This mitigates cases where outgoing messages are blocked by incoming,
+    and vice versa.
+    """
+
+    def __init__(self, protocol):
+        threading.Thread.__init__(self)
+        self.daemon = True
+        self.queue = deque()
+        self.protocol = protocol
+
+        self.cond = threading.Condition()
+        self._finished = False
+
+    def finish(self):
+        """Clear the queue and do not accept further messages."""
+        with self.cond:
+            self._finished = True
+            while len(self.queue) > 0:
+                self.queue.popleft()
+            self.cond.notify()
+
+    def push(self, msg):
+        with self.cond:
+            self.queue.append(msg)
+            self.cond.notify()
+
+    def run(self):
+        while True:
+            with self.cond:
+                if len(self.queue) == 0 and not self._finished:
+                    self.cond.wait()
+
+                if self._finished:
+                    break
+
+                msg = self.queue.popleft()
+
+            self.protocol.incoming(msg)
```

Comment:
A condition for `IncomingQueue` being useful is this call blocking when there is some amount of backpressure on the WebSocket.

Part of the reason the ROS1 branch switched to Autobahn/Twisted is it has explicit features to handle backpressure.  Tornado was infinitely queueing messages leading to OOM when a client was slow or unresponsive.  Tornado may have useful features for blocking when its outgoing buffer is above some threshold, but I did not find them.

Reply:
Nevermind, was thinking of outgoing.